src/lib/homestore_backend/hs_shard_manager.cpp

```diff
@@ -183,15 +199,33 @@ void HSHomeObject::do_shard_message_commit(int64_t lsn, ReplicationMessageHeader
     auto shard_info = deserialize_shard_info(r_cast< const char* >(value.bytes), value.size);
     switch (header.msg_type) {
     case ReplicationMessageType::CREATE_SHARD_MSG: {
-        add_new_shard_to_map(std::make_unique< HS_Shard >(shard_info, blkids.chunk_num()));
+        bool shard_exist = false;
+        {
+            std::scoped_lock lock_guard(_shard_lock);
+            shard_exist = (_shard_map.find(shard_info.id) != _shard_map.end());
+        }
+
+        if (!shard_exist) { add_new_shard_to_map(std::make_unique< HS_Shard >(shard_info, blkids.chunk_num())); }
```

Comment:
In this case we need to take the chunk out from `ChunkSelector` right?
There is a racing as we started `ChunkSelector` after metaservice recovered, which make the chunks using by  uncommitted shard looks available. Though it is probably be fine if raft statemachine will not proceed for new on_precommit).

We dont have API to mark the chunk as use

Reply:
yes, good catch for this case.  I think we need to add such api to make one chunk as using because this api will be needed by GC flow too to pick out the source/destination chunk from the heap chunkselector

Reply:
I dont know how GC flow would create the shadow shard, that part need a design doc.

but chunk_selector need to provide that API @JacksonYao287 . maybe create an issue.

Reply:
sure , we do need this API

Reply:
@JacksonYao287  the `build_per_dev_chunk_heap` was called in `on_shard_meta_blk_recover_completed`, however the raft log reply will be after meta-service recovery , all the chunk/shard/PG need to be ready before we can reply IO(like putBlob).

with that said ,the `build_per_dev_chunk_heap` has been called w/o the chunk-id of pending-replay shard.

Reply:
got your point, that is really a problem.

Reply:
about the implementation of this new API, it need to works with both cases:

1. chunk is already mark busy when write the shard header to the repldev, so when create_shard/seal_shard on_commit() is called, this chunk is already been pop out from the heap chunk selector, this API just does nothing and that is the normal case.
2. chunk is already mark busy when write the shard header to the repldev, before create_shard/seal_shard on_commit() is called, HO is restarted/crashed. after restarts,  the previous chunk will be added back to the chunk selector, so when replay on_commit() will call this new API and this new API will actually remove this chunk from chunk selector.

Reply:
@zichanglai  I am not getting your point. Chunk-selector is stateless,   it rely on shard metablks to remove in-use chunks. So in on_commit() of create_shard, if shard metadata has been updated , no need to touch chunk-selector.  Otherwise need to call the remove_chunk(chunk_id) api.

Reply:
let us assume this api is HeapChunkSelector::mark_chunk_busy(shard_id_t);   when on_commit()  is called, the associated chunk maybe is already been removed from the selector in here: https://github.com/eBay/HomeStore/blob/ddff069df74d935d10ac8cb9abc5c7ece790c8f0/src/lib/device/virtual_dev.cpp#L192.   but in recover case, this chunk is not removed from selector, so this api needed to handle both cases.

Reply:
I submit a new PR to add such interface: https://github.com/eBay/HomeObject/pull/103

Reply:
this issue is addressed in the last commit.